draft-irtf-cfrg-opaque.md

```diff
@@ -385,145 +385,329 @@ variant) of the OPRF described in {{I-D.irtf-cfrg-voprf}}.
+  
+Output:
+- secret_creds, SecretCredentials
+- export_key, an additional key
+
+Steps:
+1. prk = HKDF-Extract(envelope_nonce, Harden(y, params))
+2. auth_key = HKDF-Expand(prk, "AuthKey", Nh)
+3. export_key = HKDF-Expand(prk, "ExportKey", Nh)
+4. Create CleartextCredentials cleartext_creds with response.server_public_key and custom identifiers creds.client_identity and creds.server_identity if mode is custom_identifier
+5. expected_tag = HMAC(auth_key, concat(envelope.innver_env, cleartext_creds))
+6. If !ct_equal(envelope.auth_tag, expected_tag),
+    raise DecryptionError
+7. secret_creds = RecoverSecret(prk, envelope.inner_env)
+8. Output (secret_creds, export_key)
+~~~
+
+## Envelope Modes {#envelope-modes}
+
+The `EnvelopeMode` specifies the structure and encoding of the corresponding `InnerEnvelope`. This document specifies the values of the two aformentionned modes:
 ~~~
 enum {
-  base(1),
-  custom_identifier(2),
-  (255)
+ internal(1),
+ external(2),
+ (255)
 } EnvelopeMode;
 ~~~
 
-The `base` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+Each `EnvelopeMode` defines:
+- its own `InnerEnvelope` structure
+- the `BuildInnerEnvelope` function to build the `InnerEnvelope` structure
+- the `RecoverSecret` function to obtain the private key
 
+The inputs to HKDF-Extract and HKDF-Expand are as specified in {{RFC5869}}. The underlying hash function
+is that which is associated with the OPAQUE configuration (see {{configurations}}).
+
+<!-- TODO(Daniel):
+todo: explain rationale
+
+- this document defines opaque with 3DH
+- 3DH is over POG, i.e. can leverage internal hash-to-scalar Key gen
+- no need for the application to deal with key generation, no encrypted creds
+- for compatibility with other AKEs, external mode that allows arbitrary key kind importation with encrypted credentials.
+-->
+
+
+### Internal Key mode {#internal-mode}
+
+In this mode, the client's private and public keys are generated from a seed derived from the OPRF output. This requires the AKE to support key pair generation from a seed. The generated keys are therefore compatible with the AKE by design.
+
+For example, in the setup of 3DH using the prime-order groups defined in OPRF(reference), the secret key is generated using a HashToScalar(x) function that deterministically maps the hardened OPRF output to an element of the field, as defined in the hash to field function in {{!I-D.irtf-cfrg-hash-to-curve}}. The public key is the result of the scalar multiplication of the group's base point with the private key.
+
+With the internal key mode, the `InnerEnvelope` is constructed as follows:
 ~~~
 struct {
-  opaque client_private_key[Nsk];
-} SecretCredentials;
+ EnvelopeMode mode;
+ opaque nonce[32];
+ } InnerEnvelope;
+~~~
+mode
+: The `EnvelopeMode` value. This MUST be `internal`.
+
+nonce
+: A unique 32-byte nonce used to protect this Envelope.
+
+<!-- TODO(Daniel):
+1. Describe KeyGenFromSeed(seed) in general
```

Comment:
I'm open to suggestions on how to best formulate this.
I'm thinking of something around this

> KeyGenFromSeed(seed): expand the seed to full private/public generation in the AKE setup

Reply:
Can't this just be equal to hash_to_field with a specific DST?

Reply:
I think we should describe a generic `KeyGenFromSeed(seed)` here, and then specify that in OPAQUE-3DH we use `hash_to_field` with dst etc.. This way, it stays relevant for other AKEs, that  just need to define their own `KeyGenFromSeed(seed)` implementation.

Reply:
(Discussed offline) let's just mimic what the VOPRF draft does for HashToScalar here :)